test/e2e/util_gatewayapi_test.go

```diff
@@ -590,6 +643,132 @@ func assertGatewaySuccessful(t *testing.T, namespace, name string) (*gatewayapiv
 	return gw, nil
 }
 
+func waitForGatewayListenerCondition(t *testing.T, gatewayName types.NamespacedName, listenerName string, conditions ...metav1.Condition) error {
+	t.Helper()
+
+	gateway := &gatewayapiv1.Gateway{}
+	expected := gatewayListenerConditionMap(conditions...)
+	current := map[string]string{}
+
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 1*time.Minute, false, func(context context.Context) (bool, error) {
+		if err := kclient.Get(context, gatewayName, gateway); err != nil {
+			t.Logf("failed to get gateway %s, retrying...", gatewayName.Name)
+			return false, nil
+		}
+
+		listenerStatus := gatewayapiv1.ListenerStatus{}
+		for _, ls := range gateway.Status.Listeners {
+			if string(ls.Name) == listenerName {
+				listenerStatus = ls
+				current = gatewayListenerConditionMap(listenerStatus.Conditions...)
+				return conditionsMatchExpected(expected, current), nil
+			}
+		}
+		if &listenerStatus == nil {
+			t.Logf("listener %s not found in gateway %s, retrying...", listenerName, gatewayName.Name)
+			return false, nil
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		t.Errorf("Expected conditions: %v\n Current conditions: %v", expected, current)
+	}
+	return err
+}
+
+func gatewayListenerConditionMap(conditions ...metav1.Condition) map[string]string {
+	conds := map[string]string{}
+	for _, cond := range conditions {
+		conds[cond.Type] = string(cond.Status)
+	}
+	return conds
+}
+
+// expectedDnsRecord is used to represent a DNSRecord expectation.
+type expectedDnsRecord struct {
+	dnsName     string
+	gatewayName string
+}
+
+// assertExpectedDNSRecords polls until the DNSRecords in the default operand namespace match the given expectations.
+// The expectations parameter is a map where keys are expectations for DNSRecord and values indicate whether a DNSRecord should be present.
+func assertExpectedDNSRecords(t *testing.T, expectations map[expectedDnsRecord]bool) error {
```

Comment:
This would be easier if you did looked at one dnsRecord at a time and returned the bool for it.

Reply:
In all honesty, I can't review this as-is and it needs to be clarified asap, but doesn't block the merge unless there are continuing failures.